x/emissions/module/rewards/rewards.go

```diff
@@ -159,26 +180,52 @@ func getDistributionAndPayoutRewardsToTopicActors(
 
 func CalcTopicRewards(
 	ctx sdk.Context,
-	k keeper.Keeper,
 	weights map[uint64]*alloraMath.Dec,
 	sortedTopics []uint64,
-	sumWeight alloraMath.Dec,
-	totalReward alloraMath.Dec,
+	sumWeight alloraMath.Dec, // sum of all topic weights
+	totalReward alloraMath.Dec, // maxTotalReward in treasury
+	epochLengths map[uint64]int64, // epoch lengths for each topic
+	currentBlockEmissionDec alloraMath.Dec, // perBlockEmission
 ) (
 	map[uint64]*alloraMath.Dec,
 	error,
 ) {
+	if sumWeight.IsZero() {
+		return nil, errors.Wrapf(types.ErrInvalidReward, "total reward is zero")
+	}
+	totalTopicRewardsSum := alloraMath.ZeroDec()
 	topicRewards := make(map[TopicId]*alloraMath.Dec)
 	for _, topicId := range sortedTopics {
-		topicRewardFraction, err := GetTopicRewardFraction(weights[topicId], sumWeight)
+		topicWeight := weights[topicId]
+		if topicWeight == nil {
+			zero := alloraMath.ZeroDec()
+			topicWeight = &zero
+		}
+		topicRewardFraction, err := GetTopicRewardFraction(topicWeight, sumWeight)
 		if err != nil {
 			return nil, errors.Wrapf(err, "topic reward fraction error")
 		}
-		topicReward, err := GetTopicReward(topicRewardFraction, totalReward)
+		topicRewardPerBlock, err := GetTopicReward(topicRewardFraction, currentBlockEmissionDec)
 		if err != nil {
 			return nil, errors.Wrapf(err, "topic reward error")
 		}
-		topicRewards[topicId] = &topicReward
+		epochLength := epochLengths[topicId]
+		if epochLength <= 0 {
+			return nil, errors.Wrapf(types.ErrInvalidLengthTopic, "epoch length is nil or zero for topic %d", topicId)
+		}
+		epochLengthDec := alloraMath.NewDecFromInt64(epochLength)
+		topicRewardPerEpoch, err := topicRewardPerBlock.Mul(epochLengthDec)
+		if err != nil {
+			return nil, errors.Wrapf(err, "calcTopicRewards:topic reward multiplication with epoch length fraction error")
+		}
+		topicRewards[topicId] = &topicRewardPerEpoch
+		totalTopicRewardsSum, err = totalTopicRewardsSum.Add(topicRewardPerEpoch)
+		if err != nil {
+			return nil, errors.Wrapf(err, "calcTopicRewards: total topic rewards sum error")
+		}
+	}
+	if totalTopicRewardsSum.Gt(totalReward) {
+		return nil, errors.Wrapf(types.ErrInvalidReward, "total topic rewards sum %s is greater than total reward in treasury %s, cancelling rewards distribution", totalTopicRewardsSum.String(), totalReward.String())
```

Comment:
if we return the error with a meaningful message, I think we don't want to log it. The caller that chooses to process the error (and not propagate it) is responsible to log it.
Otherwise we end up with errors logged at every step, while we only want to have the one on top, which also writes the whole stack trace so has the whole info.